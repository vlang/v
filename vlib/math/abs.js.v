module math

fn JS.Math.abs(x f64) f64

// Returns the absolute value.
@[inline]
pub fn abs(a f64) f64 {
	return JS.Math.abs(a)
}
