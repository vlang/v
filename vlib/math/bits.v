// Copyright (c) 2019-2024 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module math

const uvnan = u64(0x7FF8000000000001)
const uvinf = u64(0x7FF0000000000000)
const uvneginf = u64(0xFFF0000000000000)
const uvone = u64(0x3FF0000000000000)
const mask = 0x7FF
const shift = 64 - 11 - 1
const bias = 1023
const normalize_smallest_mask = u64(u64(1) << 52)
const sign_mask = u64(0x8000000000000000) // (u64(1) << 63)

const frac_mask = u64((u64(1) << u64(shift)) - u64(1))

// inf returns positive infinity if sign >= 0, negative infinity if sign < 0.
pub fn inf(sign int) f64 {
	v := if sign >= 0 { math.uvinf } else { math.uvneginf }
	return f64_from_bits(v)
}

// nan returns an IEEE 754 ``not-a-number'' value.
pub fn nan() f64 {
	return f64_from_bits(math.uvnan)
}

// is_nan reports whether f is an IEEE 754 ``not-a-number'' value.
pub fn is_nan(f f64) bool {
	$if fast_math {
		if f64_bits(f) == math.uvnan {
			return true
		}
	}
	// IEEE 754 says that only NaNs satisfy f != f.
	// To avoid the floating-point hardware, could use:
	// x := f64_bits(f);
	// return u32(x>>shift)&mask == mask && x != uvinf && x != uvneginf
	return f != f
}

// is_inf reports whether f is an infinity, according to sign.
// If sign > 0, is_inf reports whether f is positive infinity.
// If sign < 0, is_inf reports whether f is negative infinity.
// If sign == 0, is_inf reports whether f is either infinity.
pub fn is_inf(f f64, sign int) bool {
	// Test for infinity by comparing against maximum float.
	// To avoid the floating-point hardware, could use:
	// x := f64_bits(f);
	// return sign >= 0 && x == uvinf || sign <= 0 && x == uvneginf;
	return (sign >= 0 && f > max_f64) || (sign <= 0 && f < -max_f64)
}

// is_finite returns true if f is finite
pub fn is_finite(f f64) bool {
	return !is_nan(f) && !is_inf(f, 0)
}

// normalize returns a normal number y and exponent exp
// satisfying x == y Ã— 2**exp. It assumes x is finite and non-zero.
pub fn normalize(x f64) (f64, int) {
	smallest_normal := 2.2250738585072014e-308 // 2**-1022
	if abs(x) < smallest_normal {
		return x * math.normalize_smallest_mask, -52
	}
	return x, 0
}
