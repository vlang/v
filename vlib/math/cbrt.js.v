module math

fn JS.Math.cbrt(x f64) f64

// cbrt calculates cubic root.
@[inline]
pub fn cbrt(a f64) f64 {
	return JS.Math.cbrt(a)
}
