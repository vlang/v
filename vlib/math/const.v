// Copyright (c) 2019-2024 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module math

pub const epsilon = 2.2204460492503130808472633361816E-16
pub const e = 2.71828182845904523536028747135266249775724709369995957496696763
pub const pi = 3.14159265358979323846264338327950288419716939937510582097494459
pub const pi_2 = pi / 2.0
pub const pi_4 = pi / 4.0
pub const phi = 1.61803398874989484820458683436563811772030917980576286213544862
pub const tau = 6.28318530717958647692528676655900576839433879875021164194988918
pub const one_over_tau = 1.0 / tau
pub const one_over_pi = 1.0 / pi
pub const tau_over2 = tau / 2.0
pub const tau_over4 = tau / 4.0
pub const tau_over8 = tau / 8.0
pub const sqrt2 = 1.41421356237309504880168872420969807856967187537694807317667974
pub const sqrt_3 = 1.73205080756887729352744634150587236694280525381038062805580697
pub const sqrt_5 = 2.23606797749978969640917366873127623544061835961152572427089724
pub const sqrt_e = 1.64872127070012814684865078781416357165377610071014801157507931
pub const sqrt_pi = 1.77245385090551602729816748334114518279754945612238712821380779
pub const sqrt_tau = 2.50662827463100050241576528481104525300698674060993831662992357
pub const sqrt_phi = 1.27201964951406896425242246173749149171560804184009624861664038
pub const ln2 = 0.693147180559945309417232121458176568075500134360255254120680009
pub const log2_e = 1.0 / ln2
pub const ln10 = 2.30258509299404568401799145468436420760110148862877297603332790
pub const log10_e = 1.0 / ln10
pub const two_thirds = 0.66666666666666666666666666666666666666666666666666666666666667

// Floating-point limit values
// max is the largest finite value representable by the type.
// smallest_non_zero is the smallest positive, non-zero value representable by the type.
pub const max_f32 = 3.40282346638528859811704183484516925440e+38 // 2**127 * (2**24 - 1) / 2**23

pub const smallest_non_zero_f32 = 1.401298464324817070923729583289916131280e-45 // 1 / 2**(127 - 1 + 23)

pub const max_f64 = 1.797693134862315708145274237317043567981e+308 // 2**1023 * (2**53 - 1) / 2**52

pub const smallest_non_zero_f64 = 4.940656458412465441765687928682213723651e-324

@[deprecated: 'use built-in constant `min_i8` instead']
@[deprecated_after: '2023-12-31']
pub const min_i8 = i8(-128)

@[deprecated: 'use built-in constant `max_i8` instead']
@[deprecated_after: '2023-12-31']
pub const max_i8 = i8(127)

@[deprecated: 'use built-in constant `min_i16` instead']
@[deprecated_after: '2023-12-31']
pub const min_i16 = i16(-32768)

@[deprecated: 'use built-in constant `max_i16` instead']
@[deprecated_after: '2023-12-31']
pub const max_i16 = i16(32767)

@[deprecated: 'use built-in constant `min_i32` instead']
@[deprecated_after: '2023-12-31']
pub const min_i32 = i32(-2147483648)

@[deprecated: 'use built-in constant `max_i32` instead']
@[deprecated_after: '2023-12-31']
pub const max_i32 = i32(2147483647)

// -9223372036854775808 is wrong, because C compilers parse literal values
// without sign first, and 9223372036854775808 overflows i64, hence the
// consecutive subtraction by 1

@[deprecated: 'use built-in constant `min_i64` instead']
@[deprecated_after: '2023-12-31']
pub const min_i64 = i64(-9223372036854775807 - 1)

@[deprecated: 'use built-in constant `max_i64` instead']
@[deprecated_after: '2023-12-31']
pub const max_i64 = i64(9223372036854775807)

@[deprecated: 'use built-in constant `min_u8` instead']
@[deprecated_after: '2023-12-31']
pub const min_u8 = u8(0)

@[deprecated: 'use built-in constant `max_u8` instead']
@[deprecated_after: '2023-12-31']
pub const max_u8 = u8(255)

@[deprecated: 'use built-in constant `min_u16` instead']
@[deprecated_after: '2023-12-31']
pub const min_u16 = u16(0)

@[deprecated: 'use built-in constant `max_u16` instead']
@[deprecated_after: '2023-12-31']
pub const max_u16 = u16(65535)

@[deprecated: 'use built-in constant `min_u32` instead']
@[deprecated_after: '2023-12-31']
pub const min_u32 = u32(0)

@[deprecated: 'use built-in constant `max_u32` instead']
@[deprecated_after: '2023-12-31']
pub const max_u32 = u32(4294967295)

@[deprecated: 'use built-in constant `min_u64` instead']
@[deprecated_after: '2023-12-31']
pub const min_u64 = u64(0)

@[deprecated: 'use built-in constant `max_u64` instead']
@[deprecated_after: '2023-12-31']
pub const max_u64 = u64(18446744073709551615)
