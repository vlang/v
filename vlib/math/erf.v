module math

/*
*                           x
 *                    2      |\
 *     erf(x)  =  ---------  | exp(-t*t)dt
 *                 sqrt(pi) \|
 *                           0
 *
 *     erfc(x) =  1-erf(x)
 *  Note that
 *              erf(-x) = -erf(x)
 *              erfc(-x) = 2 - erfc(x)
 *
 * Method:
 *      1. For |x| in [0, 0.84375]
 *          erf(x)  = x + x*R(x**2)
 *          erfc(x) = 1 - erf(x)           if x in [-.84375,0.25]
 *                  = 0.5 + ((0.5-x)-x*R)  if x in [0.25,0.84375]
 *         where R = P/Q where P is an odd poly of degree 8 and
 *         Q is an odd poly of degree 10.
 *                                               -57.90
 *                      | R - (erf(x)-x)/x | <= 2
 *
 *
 *         Remark. The formula is derived by noting
 *          erf(x) = (2/sqrt(pi))*(x - x**3/3 + x**5/10 - x**7/42 + ....)
 *         and that
 *          2/sqrt(pi) = 1.128379167095512573896158903121545171688
 *         is close to one. The interval is chosen because the fix
 *         point of erf(x) is near 0.6174 (i.e., erf(x)=x when x is
 *         near 0.6174), and by some experiment, 0.84375 is chosen to
 *         guarantee the error is less than one ulp for erf.
 *
 *      2. For |x| in [0.84375,1.25], let s_ = |x| - 1, and
 *         c = 0.84506291151 rounded to single (24 bits)
 *              erf(x)  = sign(x) * (c  + P1(s_)/Q1(s_))
 *              erfc(x) = (1-c)  - P1(s_)/Q1(s_) if x > 0
 *                        1+(c+P1(s_)/Q1(s_))    if x < 0
 *              |P1/Q1 - (erf(|x|)-c)| <= 2**-59.06
 *         Remark: here we use the taylor series expansion at x=1.
 *              erf(1+s_) = erf(1) + s_*Poly(s_)
 *                       = 0.845.. + P1(s_)/Q1(s_)
 *         That is, we use rational approximation to approximate
 *                      erf(1+s_) - (c = (single)0.84506291151)
 *         Note that |P1/Q1|< 0.078 for x in [0.84375,1.25]
 *         where
 *              P1(s_) = degree 6 poly in s_
 *              Q1(s_) = degree 6 poly in s_
 *
 *      3. For x in [1.25,1/0.35(~2.857143)],
 *              erfc(x) = (1/x)*exp(-x*x-0.5625+R1/s1)
 *              erf(x)  = 1 - erfc(x)
 *         where
 *              R1(z) = degree 7 poly in z, (z=1/x**2)
 *              s1(z) = degree 8 poly in z
 *
 *      4. For x in [1/0.35,28]
 *              erfc(x) = (1/x)*exp(-x*x-0.5625+R2/s2) if x > 0
 *                      = 2.0 - (1/x)*exp(-x*x-0.5625+R2/s2) if -6<x<0
 *                      = 2.0 - tiny            (if x <= -6)
 *              erf(x)  = sign(x)*(1.0 - erfc(x)) if x < 6, else
 *              erf(x)  = sign(x)*(1.0 - tiny)
 *         where
 *              R2(z) = degree 6 poly in z, (z=1/x**2)
 *              s2(z) = degree 7 poly in z
 *
 *      Note1:
 *         To compute exp(-x*x-0.5625+R/s), let s_ be a single
 *         precision number and s_ := x; then
 *              -x*x = -s_*s_ + (s_-x)*(s_+x)
 *              exp(-x*x-0.5626+R/s) =
 *                      exp(-s_*s_-0.5625)*exp((s_-x)*(s_+x)+R/s);
 *      Note2:
 *         Here 4 and 5 make use of the asymptotic series
 *                        exp(-x*x)
 *              erfc(x) ~ ---------- * ( 1 + Poly(1/x**2) )
 *                        x*sqrt(pi)
 *         We use rational approximation to approximate
 *              g(s_)=f(1/x**2) = log(erfc(x)*x) - x*x + 0.5625
 *         Here is the error bound for R1/s1 and R2/s2
 *              |R1/s1 - f(x)|  < 2**(-62.57)
 *              |R2/s2 - f(x)|  < 2**(-61.52)
 *
 *      5. For inf > x >= 28
 *              erf(x)  = sign(x) *(1 - tiny)  (raise inexact)
 *              erfc(x) = tiny*tiny (raise underflow) if x > 0
 *                      = 2 - tiny if x<0
 *
 *      7. special case:
 *              erf(0)  = 0, erf(inf)  = 1, erf(-inf) = -1,
 *              erfc(0) = 1, erfc(inf) = 0, erfc(-inf) = 2,
 *              erfc/erf(nan) is nan
*/
const erx = 8.45062911510467529297e-01 // 0x3FEB0AC160000000

// Coefficients for approximation to  erf in [0, 0.84375]
const efx = 1.28379167095512586316e-01 // 0x3FC06EBA8214DB69

const efx8 = 1.02703333676410069053e+00 // 0x3FF06EBA8214DB69

const pp0 = 1.28379167095512558561e-01 // 0x3FC06EBA8214DB68

const pp1 = -3.25042107247001499370e-01 // 0xBFD4CD7D691CB913

const pp2 = -2.84817495755985104766e-02 // 0xBF9D2A51DBD7194F

const pp3 = -5.77027029648944159157e-03 // 0xBF77A291236668E4

const pp4 = -2.37630166566501626084e-05 // 0xBEF8EAD6120016AC

const qq1 = 3.97917223959155352819e-01 // 0x3FD97779CDDADC09

const qq2 = 6.50222499887672944485e-02 // 0x3FB0A54C5536CEBA

const qq3 = 5.08130628187576562776e-03 // 0x3F74D022C4D36B0F

const qq4 = 1.32494738004321644526e-04 // 0x3F215DC9221C1A10

const qq5 = -3.96022827877536812320e-06 // 0xBED09C4342A26120

// Coefficients for approximation to  erf  in [0.84375, 1.25]
const pa0 = -2.36211856075265944077e-03 // 0xBF6359B8BEF77538

const pa1 = 4.14856118683748331666e-01 // 0x3FDA8D00AD92B34D

const pa2 = -3.72207876035701323847e-01 // 0xBFD7D240FBB8C3F1

const pa3 = 3.18346619901161753674e-01 // 0x3FD45FCA805120E4

const pa4 = -1.10894694282396677476e-01 // 0xBFBC63983D3E28EC

const pa5 = 3.54783043256182359371e-02 // 0x3FA22A36599795EB

const pa6 = -2.16637559486879084300e-03 // 0xBF61BF380A96073F

const qa1 = 1.06420880400844228286e-01 // 0x3FBB3E6618EEE323

const qa2 = 5.40397917702171048937e-01 // 0x3FE14AF092EB6F33

const qa3 = 7.18286544141962662868e-02 // 0x3FB2635CD99FE9A7

const qa4 = 1.26171219808761642112e-01 // 0x3FC02660E763351F

const qa5 = 1.36370839120290507362e-02 // 0x3F8BEDC26B51DD1C

const qa6 = 1.19844998467991074170e-02 // 0x3F888B545735151D

// Coefficients for approximation to  erfc in [1.25, 1/0.35]
const ra0 = -9.86494403484714822705e-03 // 0xBF843412600D6435

const ra1 = -6.93858572707181764372e-01 // 0xBFE63416E4BA7360

const ra2 = -1.05586262253232909814e+01 // 0xC0251E0441B0E726

const ra3 = -6.23753324503260060396e+01 // 0xC04F300AE4CBA38D

const ra4 = -1.62396669462573470355e+02 // 0xC0644CB184282266

const ra5 = -1.84605092906711035994e+02 // 0xC067135CEBCCABB2

const ra6 = -8.12874355063065934246e+01 // 0xC054526557E4D2F2

const ra7 = -9.81432934416914548592e+00 // 0xC023A0EFC69AC25C

const sa1 = 1.96512716674392571292e+01 // 0x4033A6B9BD707687

const sa2 = 1.37657754143519042600e+02 // 0x4061350C526AE721

const sa3 = 4.34565877475229228821e+02 // 0x407B290DD58A1A71

const sa4 = 6.45387271733267880336e+02 // 0x40842B1921EC2868

const sa5 = 4.29008140027567833386e+02 // 0x407AD02157700314

const sa6 = 1.08635005541779435134e+02 // 0x405B28A3EE48AE2C

const sa7 = 6.57024977031928170135e+00 // 0x401A47EF8E484A93

const sa8 = -6.04244152148580987438e-02 // 0xBFAEEFF2EE749A62

// Coefficients for approximation to  erfc in [1/.35, 28]
const rb0 = -9.86494292470009928597e-03 // 0xBF84341239E86F4A

const rb1 = -7.99283237680523006574e-01 // 0xBFE993BA70C285DE

const rb2 = -1.77579549177547519889e+01 // 0xC031C209555F995A

const rb3 = -1.60636384855821916062e+02 // 0xC064145D43C5ED98

const rb4 = -6.37566443368389627722e+02 // 0xC083EC881375F228

const rb5 = -1.02509513161107724954e+03 // 0xC09004616A2E5992

const rb6 = -4.83519191608651397019e+02 // 0xC07E384E9BDC383F

const sb1 = 3.03380607434824582924e+01 // 0x403E568B261D5190

const sb2 = 3.25792512996573918826e+02 // 0x40745CAE221B9F0A

const sb3 = 1.53672958608443695994e+03 // 0x409802EB189D5118

const sb4 = 3.19985821950859553908e+03 // 0x40A8FFB7688C246A

const sb5 = 2.55305040643316442583e+03 // 0x40A3F219CEDF3BE6

const sb6 = 4.74528541206955367215e+02 // 0x407DA874E79FE763

const sb7 = -2.24409524465858183362e+01

// erf returns the error function of x.
//
// special cases are:
// erf(+inf) = 1
// erf(-inf) = -1
// erf(nan) = nan
pub fn erf(a f64) f64 {
	mut x := a
	very_tiny := 2.848094538889218e-306 // 0x0080000000000000
	small_ := 1.0 / f64(u64(1) << 28) // 2**-28
	if is_nan(x) {
		return nan()
	}
	if is_inf(x, 1) {
		return 1.0
	}
	if is_inf(x, -1) {
		return f64(-1)
	}
	mut sign := false
	if x < 0 {
		x = -x
		sign = true
	}
	if x < 0.84375 { // |x| < 0.84375
		mut temp := 0.0
		if x < small_ { // |x| < 2**-28
			if x < very_tiny {
				temp = 0.125 * (8.0 * x + efx8 * x) // avoid underflow
			} else {
				temp = x + efx * x
			}
		} else {
			z := x * x
			r := pp0 + z * (pp1 + z * (pp2 + z * (pp3 + z * pp4)))
			s_ := 1.0 + z * (qq1 + z * (qq2 + z * (qq3 + z * (qq4 + z * qq5))))
			y := r / s_
			temp = x + x * y
		}
		if sign {
			return -temp
		}
		return temp
	}
	if x < 1.25 { // 0.84375 <= |x| < 1.25
		s_ := x - 1
		p := pa0 + s_ * (pa1 + s_ * (pa2 + s_ * (pa3 + s_ * (pa4 + s_ * (pa5 + s_ * pa6)))))
		q := 1.0 + s_ * (qa1 + s_ * (qa2 + s_ * (qa3 + s_ * (qa4 + s_ * (qa5 + s_ * qa6)))))
		if sign {
			return -erx - p / q
		}
		return erx + p / q
	}
	if x >= 6 { // inf > |x| >= 6
		if sign {
			return -1
		}
		return 1.0
	}
	s_ := 1.0 / (x * x)
	mut r := 0.0
	mut s := 0.0
	if x < 1.0 / 0.35 { // |x| < 1 / 0.35  ~ 2.857143
		r = ra0 + s_ * (ra1 + s_ * (ra2 + s_ * (ra3 + s_ * (ra4 + s_ * (ra5 + s_ * (ra6 +
			s_ * ra7))))))
		s = 1.0 + s_ * (sa1 + s_ * (sa2 + s_ * (sa3 + s_ * (sa4 + s_ * (sa5 + s_ * (sa6 +
			s_ * (sa7 + s_ * sa8)))))))
	} else { // |x| >= 1 / 0.35  ~ 2.857143
		r = rb0 + s_ * (rb1 + s_ * (rb2 + s_ * (rb3 + s_ * (rb4 + s_ * (rb5 + s_ * rb6)))))
		s = 1.0 + s_ * (sb1 + s_ * (sb2 + s_ * (sb3 + s_ * (sb4 + s_ * (sb5 + s_ * (sb6 +
			s_ * sb7))))))
	}
	z := f64_from_bits(f64_bits(x) & 0xffffffff00000000) // pseudo-single (20-bit) precision x
	r_ := exp(-z * z - 0.5625) * exp((z - x) * (z + x) + r / s)
	if sign {
		return r_ / x - 1.0
	}
	return 1.0 - r_ / x
}

// erfc returns the complementary error function of x.
//
// special cases are:
// erfc(+inf) = 0
// erfc(-inf) = 2
// erfc(nan) = nan
pub fn erfc(a f64) f64 {
	mut x := a
	tiny := 1.0 / f64(u64(1) << 56) // 2**-56
	// special cases
	if is_nan(x) {
		return nan()
	}
	if is_inf(x, 1) {
		return 0.0
	}
	if is_inf(x, -1) {
		return 2.0
	}
	mut sign := false
	if x < 0 {
		x = -x
		sign = true
	}
	if x < 0.84375 { // |x| < 0.84375
		mut temp := 0.0
		if x < tiny { // |x| < 2**-56
			temp = x
		} else {
			z := x * x
			r := pp0 + z * (pp1 + z * (pp2 + z * (pp3 + z * pp4)))
			s_ := 1.0 + z * (qq1 + z * (qq2 + z * (qq3 + z * (qq4 + z * qq5))))
			y := r / s_
			if x < 0.25 { // |x| < 1.0/4
				temp = x + x * y
			} else {
				temp = 0.5 + (x * y + (x - 0.5))
			}
		}
		if sign {
			return 1.0 + temp
		}
		return 1.0 - temp
	}
	if x < 1.25 { // 0.84375 <= |x| < 1.25
		s_ := x - 1
		p := pa0 + s_ * (pa1 + s_ * (pa2 + s_ * (pa3 + s_ * (pa4 + s_ * (pa5 + s_ * pa6)))))
		q := 1.0 + s_ * (qa1 + s_ * (qa2 + s_ * (qa3 + s_ * (qa4 + s_ * (qa5 + s_ * qa6)))))
		if sign {
			return 1.0 + erx + p / q
		}
		return 1.0 - erx - p / q
	}
	if x < 28 { // |x| < 28
		s_ := 1.0 / (x * x)
		mut r := 0.0
		mut s := 0.0
		if x < 1.0 / 0.35 { // |x| < 1 / 0.35 ~ 2.857143
			r = ra0 + s_ * (ra1 + s_ * (ra2 + s_ * (ra3 + s_ * (ra4 + s_ * (ra5 + s_ * (ra6 +
				s_ * ra7))))))
			s = 1.0 + s_ * (sa1 + s_ * (sa2 + s_ * (sa3 + s_ * (sa4 + s_ * (sa5 + s_ * (sa6 +
				s_ * (sa7 + s_ * sa8)))))))
		} else { // |x| >= 1 / 0.35 ~ 2.857143
			if sign && x > 6 {
				return 2.0 // x < -6
			}
			r = rb0 + s_ * (rb1 + s_ * (rb2 + s_ * (rb3 + s_ * (rb4 + s_ * (rb5 + s_ * rb6)))))
			s = 1.0 + s_ * (sb1 + s_ * (sb2 + s_ * (sb3 + s_ * (sb4 + s_ * (sb5 + s_ * (sb6 +
				s_ * sb7))))))
		}
		z := f64_from_bits(f64_bits(x) & 0xffffffff00000000) // pseudo-single (20-bit) precision x
		r_ := exp(-z * z - 0.5625) * exp((z - x) * (z + x) + r / s)
		if sign {
			return 2.0 - r_ / x
		}
		return r_ / x
	}
	if sign {
		return 2.0
	}
	return 0.0
}
