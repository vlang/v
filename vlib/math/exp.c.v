module math

fn C.exp(x f64) f64

fn C.exp2(x f64) f64

fn C.ldexp(x f64, exp int) f64

// exp returns e**x, the base-e exponential of x.
@[inline]
pub fn exp(x f64) f64 {
	return C.exp(x)
}

// exp2 returns 2**x, the base-2 exponential of x.
@[inline]
pub fn exp2(x f64) f64 {
	return C.exp2(x)
}

// ldexp calculates frac*(2**exp).
@[inline]
pub fn ldexp(frac f64, exp int) f64 {
	return C.ldexp(frac, exp)
}
