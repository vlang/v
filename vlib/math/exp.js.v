module math

fn JS.Math.exp(x f64) f64

// exp calculates exponent of the number (math.pow(math.E, x)).
@[inline]
pub fn exp(x f64) f64 {
	mut res := 0.0
	#res.val = Math.exp(x)

	return res
}
