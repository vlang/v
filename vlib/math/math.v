// Copyright (c) 2019-2024 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module math

// aprox_sin returns an approximation of sin(a) made using lolremez
pub fn aprox_sin(a f64) f64 {
	a0 := 1.91059300966915117e-31
	a1 := 1.00086760103908896
	a2 := -1.21276126894734565e-2
	a3 := -1.38078780785773762e-1
	a4 := -2.67353392911981221e-2
	a5 := 2.08026600266304389e-2
	a6 := -3.03996055049204407e-3
	a7 := 1.38235642404333740e-4
	return a0 + a * (a1 + a * (a2 + a * (a3 + a * (a4 + a * (a5 + a * (a6 + a * a7))))))
}

// aprox_cos returns an approximation of cos(a) made using lolremez
pub fn aprox_cos(a f64) f64 {
	a0 := 9.9995999154986614e-1
	a1 := 1.2548995793001028e-3
	a2 := -5.0648546280678015e-1
	a3 := 1.2942246466519995e-2
	a4 := 2.8668384702547972e-2
	a5 := 7.3726485210586547e-3
	a6 := -3.8510875386947414e-3
	a7 := 4.7196604604366623e-4
	a8 := -1.8776444013090451e-5
	return a0 + a * (a1 + a * (a2 + a * (a3 + a * (a4 + a * (a5 + a * (a6 + a * (a7 + a * a8)))))))
}

// copysign returns a value with the magnitude of x and the sign of y
@[inline]
pub fn copysign(x f64, y f64) f64 {
	return f64_from_bits((f64_bits(x) & ~sign_mask) | (f64_bits(y) & sign_mask))
}

// degrees converts an angle in radians to a corresponding angle in degrees.
@[inline]
pub fn degrees(radians f64) f64 {
	return radians * (180.0 / pi)
}

// angle_diff calculates the difference between angles in radians
@[inline]
pub fn angle_diff(radian_a f64, radian_b f64) f64 {
	mut delta := fmod(radian_b - radian_a, tau)
	delta = fmod(delta + 1.5 * tau, tau)
	delta -= .5 * tau
	return delta
}

@[params]
pub struct DigitParams {
pub:
	base    int = 10
	reverse bool
}

// digits returns an array of the digits of `num` in the given optional `base`.
// The `num` argument accepts any integer type (i8|i16|int|isize|i64), and will be cast to i64
// The `base:` argument is optional, it will default to base: 10.
// This function returns an array of the digits in reverse order i.e.:
// Example: assert math.digits(12345, base: 10) == [5,4,3,2,1]
// You can also use it, with an explicit `reverse: true` parameter,
// (it will do a reverse of the result array internally => slower):
// Example: assert math.digits(12345, reverse: true) == [1,2,3,4,5]
pub fn digits(num i64, params DigitParams) []int {
	// set base to 10 initially and change only if base is explicitly set.
	mut b := params.base
	if b < 2 {
		panic('digits: Cannot find digits of n with base ${b}')
	}
	mut n := num
	mut sign := 1
	if n < 0 {
		sign = -1
		n = -n
	}

	mut res := []int{}
	if n == 0 {
		// short-circuit and return 0
		res << 0
		return res
	}
	for n != 0 {
		next_n := n / b
		res << int(n - next_n * b)
		n = next_n
	}

	if sign == -1 {
		res[res.len - 1] *= sign
	}

	if params.reverse {
		res = res.reverse()
	}

	return res
}

// count_digits return the number of digits in the number passed.
// Number argument accepts any integer type (i8|i16|int|isize|i64) and will be cast to i64
pub fn count_digits(number i64) int {
	mut n := number
	if n == 0 {
		return 1
	}
	mut c := 0
	for n != 0 {
		n = n / 10
		c++
	}
	return c
}

// minmax returns the minimum and maximum value of the two provided.
pub fn minmax(a f64, b f64) (f64, f64) {
	if a < b {
		return a, b
	}
	return b, a
}

// clamp returns x constrained between a and b
@[inline]
pub fn clamp(x f64, a f64, b f64) f64 {
	if x < a {
		return a
	}
	if x > b {
		return b
	}
	return x
}

// sign returns the corresponding sign -1.0, 1.0 of the provided number.
// if n is not a number, its sign is nan too.
@[inline]
pub fn sign(n f64) f64 {
	// dump(n)
	if is_nan(n) {
		return nan()
	}
	return copysign(1.0, n)
}

// signi returns the corresponding sign -1, 1 of the provided number.
@[inline]
pub fn signi(n f64) int {
	return int(copysign(1.0, n))
}

// radians converts an angle in degrees to a corresponding angle in radians.
@[inline]
pub fn radians(degrees f64) f64 {
	return degrees * (pi / 180.0)
}

// signbit returns a value with the boolean representation of the sign for x
@[inline]
pub fn signbit(x f64) bool {
	return f64_bits(x) & sign_mask != 0
}

// tolerance checks if a and b difference are less than or equal to the tolerance value
pub fn tolerance(a f64, b f64, tol f64) bool {
	mut ee := tol
	// Multiplying by ee here can underflow denormal values to zero.
	// Check a==b so that at least if a and b are small and identical
	// we say they match.
	if a == b {
		return true
	}
	mut d := a - b
	if d < 0 {
		d = -d
	}
	// note: b is correct (expected) value, a is actual value.
	// make error tolerance a fraction of b, not a.
	if b != 0 {
		ee = ee * b
		if ee < 0 {
			ee = -ee
		}
	}
	return d < ee
}

// close checks if a and b are within 1e-14 of each other
pub fn close(a f64, b f64) bool {
	return tolerance(a, b, 1e-14)
}

// veryclose checks if a and b are within 4e-16 of each other
pub fn veryclose(a f64, b f64) bool {
	return tolerance(a, b, 4e-16)
}

// alike checks if a and b are equal
pub fn alike(a f64, b f64) bool {
	// eprintln('>>> a: ${f64_bits(a):20} | b: ${f64_bits(b):20} | a==b: ${a == b} | a: ${a:10} | b: ${b:10}')
	// compare a and b, ignoring their last 2 bits:
	if f64_bits(a) & 0xFFFF_FFFF_FFFF_FFFC == f64_bits(b) & 0xFFFF_FFFF_FFFF_FFFC {
		return true
	}
	if a == -0 && b == 0 {
		return true
	}
	if a == 0 && b == -0 {
		return true
	}
	if is_nan(a) && is_nan(b) {
		return true
	}
	if a == b {
		return signbit(a) == signbit(b)
	}
	return false
}

fn is_odd_int(x f64) bool {
	xi, xf := modf(x)
	return xf == 0 && (i64(xi) & 1) == 1
}

fn is_neg_int(x f64) bool {
	if x < 0 {
		_, xf := modf(x)
		return xf == 0
	}
	return false
}
