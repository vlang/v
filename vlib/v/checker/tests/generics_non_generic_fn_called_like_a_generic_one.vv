import math

fn main() {
	x := math.sin[f64](1.0)
	println(x)
}
