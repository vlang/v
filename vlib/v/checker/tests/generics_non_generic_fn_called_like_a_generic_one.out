vlib/v/checker/tests/generics_non_generic_fn_called_like_a_generic_one.vv:4:15: error: a non generic function called like a generic one
    2 | 
    3 | fn main() {
    4 |     x := math.sin[f64](1.0)
      |                  ~~~~~
    5 |     println(x)
    6 | }
